src/full/Agda/TypeChecking/Rules/Application.hs

```diff
@@ -1044,23 +1044,47 @@ inferOrCheckProjApp e o ds args mt = do
         [ "  principal arg " <+> prettyTCM arg
         , "  has type "      <+> prettyTCM ta
         ]
-      inferOrCheckProjAppToKnownPrincipalArg e o ds args mt k v0 ta
+      inferOrCheckProjAppToKnownPrincipalArg e o ds args mt k v0 ta Nothing
 
 -- | Same arguments 'inferOrCheckProjApp' above but also gets the position,
 --   value and type of the principal argument.
-inferOrCheckProjAppToKnownPrincipalArg ::
-  A.Expr -> ProjOrigin -> List1 QName -> A.Args -> Maybe (Comparison, Type) ->
-  Int -> Term -> Type -> TCM (Term, Type, CheckedTarget)
-inferOrCheckProjAppToKnownPrincipalArg e o ds args mt k v0 ta = do
+inferOrCheckProjAppToKnownPrincipalArg
+  :: A.Expr
+     -- ^ The whole expression which constitutes the application.
+  -> ProjOrigin
+     -- ^ The origin of the projection involved in this projection application.
+  -> List1 QName
+     -- ^ The projection name (potentially ambiguous).
+  -> A.Args
+     -- ^ The arguments to the projection.
+  -> Maybe (Comparison, Type)
+     -- ^ The expected type of the expression (if 'Nothing', infer it).
+  -> Int
+     -- ^ The position of the principal argument.
+  -> Term
+     -- ^ The value of the principal argument.
+  -> Type
+     -- ^ The type of the principal argument.
+  -> Maybe PrincipalArgTypeMetas
+     -- ^ The metas previously created for the principal argument's type, when
+     --   picking up a postponed problem. 'Nothing', otherwise.
+  -> TCM (Term, Type, CheckedTarget)
+     -- ^ The type-checked expression and its type (if successful).
+inferOrCheckProjAppToKnownPrincipalArg e o ds args mt k v0 ta mpatm = do
   let cmp = caseMaybe mt CmpEq fst
-      postpone b = do
+      postpone b vargs ta = do
```

Comment:
Maybe I shouldn't bother you further, but the logical solution would be that `postpone` gets the `PrincipalArgTypeMetas` pair as argument.

Reply:
Oh, good point. Yes. As I meant to indicate that "these two things go together", I should also do that in the arguments of `postpone`. That's the whole point of `PrincipalArgTypeMetas`, so why not use it everywhere, where "these two things" get passed around. I'll make the change.